src/main/host/syscall/handler/epoll.rs

```diff
@@ -0,0 +1,423 @@
+
+        // We'll need to look up descriptors.
+        let desc_table = ctx.objs.thread.descriptor_table_borrow(ctx.objs.host);
+
+        // Get the epoll descriptor, or return early if it doesn't exist.
+        let epoll = {
+            let desc = Self::get_descriptor(&desc_table, epfd)?;
+
+            let CompatFile::New(epoll) = desc.file() else {
+                return Err(Errno::EINVAL.into());
+            };
+            let File::Epoll(epoll) = epoll.inner_file() else {
+                return Err(Errno::EINVAL.into());
+            };
+
+            epoll
+        };
+
+        // Get the target descriptor, or return errors as appropriate.
+        let target = {
+            let desc = Self::get_descriptor(&desc_table, fd)?;
+
+            // Our epoll implementation only supports adding new Rust descriptor types.
+            // However, the only legacy type remaining in Shadow is a regular file, and
+            // epoll_ctl(2) states that EPERM should be returned for regular files and
+            // other files that don't support epolling.
+            match desc.file() {
+                CompatFile::New(file) => file.inner_file().clone(),
+                CompatFile::Legacy(file) => {
+                    let file_type = unsafe { cshadow::legacyfile_getType(file.ptr()) };
+                    if file_type == cshadow::_LegacyFileType_DT_FILE {
+                        // Epoll doesn't support regular files.
+                        return Err(Errno::EPERM.into());
+                    } else {
+                        // Our implementation doesn't support other legacy types.
+                        // We don't think we have such types remaining, but warn anyway.
+                        warn_once_then_trace!(
+                            "(LOG_ONCE) Attempted to add a legacy file to an \
+                            epoll file, which shadow doesn't support"
+                        );
+                        return Err(Errno::EINVAL.into());
+                    }
+                }
+            }
+        };
+
+        // An epoll instance is not allowed to monitor itself.
+        if epfd == fd {
+            return Err(Errno::EINVAL.into());
+        }
+
+        // Extract the operation.
+        let Ok(op) = EpollCtlOp::try_from(op) else {
+            log::debug!("Invalid epoll op: {op}");
+            return Err(Errno::EINVAL.into());
+        };
+
+        // Extract the events and data.
+        let (events, data) = if op == EpollCtlOp::EPOLL_CTL_DEL {
+            // epoll_ctl(2): Since Linux 2.6.9, the event pointer is ignored and can be specified as
+            // NULL when using EPOLL_CTL_DEL.
+            (EpollEvents::empty(), 0)
+        } else {
+            let mem = ctx.objs.process.memory_borrow();
+            let ev = mem.read(event_ptr)?;
+
+            let Some(mut events) = EpollEvents::from_bits(ev.events) else {
+                log::debug!("Invalid epoll_ctl events: {}", { ev.events });
```

Comment:
nit: the braces can be removed

Reply:
I'll add a comment to the code to reference `rustc --explain E0793`.

Reply:
Ah I never realized this is a packed struct on x86-64.